internal/cas/oci.go

```diff
@@ -0,0 +1,176 @@
+	}
+
+	content, err := os.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("%s: %s: %w", target.Digest, target.MediaType, errdef.ErrNotFound)
+		}
+		return nil, err
+	}
+
+	return io.NopCloser(bytes.NewReader(content)), nil
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (o *OCI) Push(_ context.Context, expected ocispec.Descriptor, content io.Reader) error {
+	// calculate the target path for the content.
+	target, err := o.blobPath(expected.Digest)
+	if err != nil {
+		return fmt.Errorf("%s: %s: %w", expected.Digest, expected.MediaType, errdef.ErrInvalidDigest)
+	}
+
+	// check if the target content exists in the blob directory.
+	if _, err := os.Stat(target); err == nil {
+		return fmt.Errorf("%s: %s: %w", expected.Digest, expected.MediaType, errdef.ErrAlreadyExists)
+	}
+
+	// write the content to a temporary ingest file.
+	ingest, err := o.ingest(expected, content)
+	if err != nil {
+		return err
+	}
+
+	if err := ensureDir(filepath.Dir(target)); err != nil {
+		return err
+	}
+
+	// move the content from the temporary ingest file to the target path.
+	if err := os.Rename(ingest, target); err != nil {
+		// clean up ingest file in case of error
+		if err := os.Remove(ingest); err != nil {
+			return fmt.Errorf("failed to clean up ingest file: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// Exists returns true if the described content exists.
+func (o *OCI) Exists(_ context.Context, target ocispec.Descriptor) (bool, error) {
+	path, err := o.blobPath(target.Digest)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+
+	return false, err
+}
+
+// blobPath calculates blob path from the given digest.
+func (o *OCI) blobPath(dgst digest.Digest) (string, error) {
+	if err := dgst.Validate(); err != nil {
+		return "", fmt.Errorf("cannot calculate blob path from invalid digest %s: %v", dgst.String(), err)
+	}
+
+	return filepath.Join(o.blobRoot, dgst.Algorithm().String(), dgst.Encoded()), nil
+}
+
+// ingest write the content into a temporary ingest file.
+func (o *OCI) ingest(expected ocispec.Descriptor, content io.Reader) (string, error) {
+	if err := ensureDir(o.ingestRoot); err != nil {
+		return "", err
+	}
+
+	// create a temp file with the file name format "blobDigest_randomString"
+	// in the ingest direcotry.
+	// Go ensures that multiple programs or goroutines calling CreateTemp
+	// simultaneously will not choose the same file.
+	f, err := os.CreateTemp(o.ingestRoot, expected.Digest.Encoded()+"_*")
+	if err != nil {
+		return "", fmt.Errorf("failed to create ingest file: %v", err)
+	}
+	defer f.Close()
+
+	value, err := ioutil.ReadAll(content, expected)
+	if err != nil {
+		return "", fmt.Errorf("failed to read from content: %v", err)
+	}
+	if _, err := f.Write(value); err != nil {
+		return "", fmt.Errorf("failed to write ingest file: %v", err)
+	}

```

Comment:
You may take a look at [io.Copy](https://pkg.go.dev/io#Copy), or its optimized version [io.CopyBuffer](https://pkg.go.dev/io#CopyBuffer). If you choose to use `io.CopyBuffer`, [sync.Pool](https://pkg.go.dev/sync#Pool) is your friend.